src/validation/read_constant_expression.rs

```diff
@@ -13,27 +99,42 @@ pub(super) fn read_constant_expression(wasm: &mut WasmReader) -> Result<Span> {
 
         use crate::core::reader::types::opcode::*;
         match first_instr_byte {
+            // Missing: ref.null, ref.func, global.get
             END => {
+                // The stack must only contain the global's valtype
+                if this_global_valtype.is_some() {
+                    stack.assert_val_types(&[this_global_valtype.unwrap()])?;
+                }
                 return Ok(Span::new(start_pc, wasm.pc - start_pc + 1));
             }
-            GLOBAL_GET => {
-                wasm.read_var_u32()?;
-            }
             I32_CONST => {
-                wasm.read_var_i32()?;
+                let _num = wasm.read_var_i32()?;
+                stack.push_valtype(ValType::NumType(NumType::I32));
             }
             I64_CONST => {
-                wasm.read_var_i64()?;
+                let _num = wasm.read_var_i64()?;
+                stack.push_valtype(ValType::NumType(NumType::I64));
             }
-            I32_ADD | I32_SUB | I32_MUL => {}
-            I64_ADD | I64_SUB | I64_MUL => {}
-            REF_NULL => {
-                RefType::read(wasm).unwrap();
-            }
-            REF_FUNC => {
-                wasm.read_var_u32().unwrap();
+            I32_ADD | I32_SUB | I32_MUL => {
+                stack.assert_pop_val_type(ValType::NumType(NumType::I32))?;
+                stack.assert_pop_val_type(ValType::NumType(NumType::I32))?;
+
+                stack.push_valtype(ValType::NumType(NumType::I32));
             }
+            I64_ADD | I64_SUB | I64_MUL => {
+                stack.assert_pop_val_type(ValType::NumType(NumType::I64))?;
+                stack.assert_pop_val_type(ValType::NumType(NumType::I64))?;
 
+                stack.push_valtype(ValType::NumType(NumType::I64));
+            }
+            // REF_NULL => {
+            //     RefType::read(wasm).unwrap();
+                
+            //     stack.push_valtype(ValType::RefType(RefType::));
+            // }
+            // REF_FUNC => {
+            //     wasm.read_var_u32().unwrap();
+            // }
```

Comment:
We can have the op codes, but have the implementations as todo if there are issues here